### ISSUE TITLE
runc with user namespace enabled fails to bind mount host dirs with 750 permission

### ISSUE THREAD

**Issue by alban** [opencontainers/runc#2484]
When config.json has the following:
- user namespace enabled
- a bind mount from a host directory that is a sub-directory of one with with `drwxr-x---` permission, it fails with the error message:

```
# time="2020-06-22T13:48:26Z" level=error msg="container_linux.go:367:
starting container process caused: process_linux.go:459:
container init caused: rootfs_linux.go:58:
mounting \"/tmp/busyboxtest/source-inaccessible/dir\"
to rootfs at \"/tmp/inaccessible\" caused:
stat /tmp/busyboxtest/source-inaccessible/dir: permission denied"
```

I implemented a reproducer in the integration tests, along with the explanation and a workaround when started from a systemd unit: https://github.com/opencontainers/runc/pull/2483

**Comment by cyphar** [comment]
This is expected behaviour. Even if we were to blindly `mount(2)` the path, path lookup would fail because by the time we are mounting filesystems the container is running in the user namespace and thus `CAP_DAC_OVERRIDE` doesn't work anymore.

There are possible ways to work around this (such as looking up the path while running with privileges on the host, then bind-mounting those file descriptors once the container has started) but that will make our mounting code much more complicated. Is this something that you really need, or is it possible to adjust your system configuration?

**Comment by alban** [comment]
> Is this something that you really need, or is it possible to adjust your system configuration?

I think that would be nice to have: `/var/lib/kubelet` contains cryptographic keys, pods directories with ConfigMap and Secrets. Currently, only root can access it. It feels wrong if we give access to all users, even if it is only traversal rights (`--x`).

I am not sure what is the best workaround. I have not thought of passing file descriptors.

- Open file descriptors in parent privileged process, mounting in the child process
  - Pros: should work for all corner cases
  - Cons(?): does it require the new Linux mount API? Does it work with old kernels?
  - Cons: more complicated code
- Do the mounts in the child mntns but in the host userns (so we retain privileges).
  - Cons: not possible in Golang: Linux does not allow `setns()` on mntns with multithreaded programs. And Linux does not allow to go back to the parent userns with `setns()`.
  - Cons: needs complicated interprocess synchronisation
- runc adding gid=0 as supplemental group
  - Cons: difficult to do with Golang because `syscall.Setgroups()` is per thread. [POSIX setgroups() support](https://go-review.googlesource.com/c/go/+/210639/) is in development but not there yet.
  - Cons: gid=0 might not be the only gid to add in supplemental groups: if a directory belongs to another group than root, that other group would need to be added too.
  - Cons: if a directory is `drwx------`, then adding a supplemental user would not work. There is no concept of supplemental user.
  - Pros: maybe we could use [LockOSThread](https://golang.org/pkg/runtime/#LockOSThread) + [syscall.Setgroups](https://golang.org/pkg/syscall/#Setgroups) just before starting the new process.
- Documenting that processes starting `runc` (such as containerd) should run with systemd option `SupplementaryGroups=0`.
  - Pros: easy to implement
  - Pros: works good enough for the `/var/lib/kubelet` use case
  - Cons: it is implementing a workaround in the wrong place in my opinion
  - Cons: would not work with more complicated use cases such as a directory with `drwx------` or belonging to another group than root (but I don't have a use case that would require it at the moment)

**Comment by cyphar** [comment]
> Open file descriptors in parent privileged process, mounting in the child process
>  * Cons(?): does it require the new Linux mount API? Does it work with old kernels?

Yes, it'll work with basically all old kernels I can think of. While the mount API does make the fact you can bind-mount file descriptors explicit, you've always been able to do it through `/proc/self/fd/...`. Whether this is a bug or a feature is up to your world-view (it's definitely a bug though).

Honestly, given all of the other options I think it's the least-ugly and most-correct solution. :shrug:

**Comment by cyphar** [comment]
It actually shouldn't be *too* complicated. We can just rewrite the configuration we pass to the container to use `/proc/self/fd/...`. The only issue is that the mountpoint fds can't be marked `O_CLOEXEC` and so we'd need to make sure we manually close them (though we already have code to do this at the very last stage of `runc init`).

**Comment by alban** [comment]
Due to the kernel cross-mntns check, I tried this approach: open the mount sources in the host userns but in the container mntns, and pass the fds by SCM_RIGHTS. It works on my test scenario.
https://github.com/opencontainers/runc/pull/2576
